File: ConsoleApplication/Tree/BottomView.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApplication1.Tree
{
    public class BottomView
    {
        public void PrintBottomView(TreeNodeWithHd root)
        {
            Queue<TreeNodeWithHd> q = new Queue<TreeNodeWithHd>();
            root.Hd = 0;
            q.Enqueue(root);

            SortedDictionary<int, int> sd = new SortedDictionary<int, int>();
            sd.Add(root.Hd, root.Data);

            while (q.Count != 0)
            {
                TreeNodeWithHd node = q.Dequeue();

                if (!sd.ContainsKey(node.Hd))
                {
                    sd.Add(node.Hd, node.Data);
                }

                if (node.Left != null)
                {
                    TreeNodeWithHd nodeWithHdL = new TreeNodeWithHd(node.Left);
                    nodeWithHdL.Hd = node.Hd - 1;
                    q.Enqueue(nodeWithHdL);
                }

                if (node.Right != null)
                {
                    TreeNodeWithHd nodeWithHdR = new TreeNodeWithHd(node.Left);
                    nodeWithHdR.Hd = node.Hd + 1;
                    q.Enqueue(nodeWithHdR);
                }
            }

            foreach(int data in sd.Values)
            {
                Console.Write($"{data};");
            }

        }
    }
}

File: Tasks/src/trees/BottomView.kt
package trees

import java.util.*
import kotlin.collections.HashMap

class BottomView {
    fun find() {
        val root: TreeNodeWithHd = Tree.getExampleTree3()
        val queue: Queue<TreeNodeWithHd> = ArrayDeque<TreeNodeWithHd>()
        val map = hashMapOf<Int, TreeNodeWithHd>()
        findBorromView(root, queue, map)

        map.toSortedMap().values.forEach {
            print("${it.value}, ")
        }
    }

    private fun findBorromView(root: TreeNodeWithHd, queue: Queue<TreeNodeWithHd>, map: HashMap<Int, TreeNodeWithHd>) {
        queue.add(root)

        while(queue.isNotEmpty()) {
            val node = queue.remove()
            map[node.hd] = node

            if (node.left != null) {
                node.left?.hd = node.hd - 1
                queue.add(node.left)
            }

            if (node.right != null) {
                node.right?.hd = node.hd + 1
                queue.add(node.right)
            }
        }
    }

}